### DISCUSSION TITLE
ESP32-S3 - BLE MIDI controller | Help with code

### DISCUSSION THREAD

**Discussion by nteflgrnt** [FortySevenEffects/arduino_midi_library#356] [General]
Hi There,

I'm building a bluetooth midi controller using my ESP32-S3 board- spoiler alert, I am a hobbyist so my knowledge is limited at best. 

The controller I'm building needs to send midi signal from 2x buttons, 2x B10K potentiometers and 2x rotary encoders. So far I have managed to get the bluetooth midi connection working fine and the 2 buttons are sending midi over bluetooth successfully, but I'm completely at a loss on how to code the pots and the rotary encoders. Could anyone assist me with this?

BOARD USED: ESP32-S3

POT1 PIN: 4
POT2 PIN: 5

BUTTON1 PIN: 6
BUTTON2 PIN: 7

ROTARY1a PIN: 15
ROTARY1b PIN: 16

ROTARY2a PIN: 17
ROTARY2b PIN: 18

Any help would be massively appreciated! 

Here's my current code that includes the 2x buttons:

`#include <BLEMIDI_Transport.h>
#include <hardware/BLEMIDI_ESP32.h>

//BLEMIDI_CREATE_DEFAULT_INSTANCE()
 BLEMIDI_CREATE_INSTANCE("FAM_INSTRUMENT", MIDI)

const int btnPin1 = 6;
const int btnPin2 = 7;

int midiNote;
int btnState;
int lastBtnState;

void setup() {
  pinMode(btnPin1, INPUT_PULLUP);
  pinMode(btnPin2, INPUT_PULLUP);
  
  MIDI.begin();

}

void loop() {
  
  lastBtnState = btnState;
  btnState = digitalRead(btnPin1);

  if(lastBtnState == 1 && btnState == 0){
    MIDI.sendNoteOn(60, 127, 1);
    delay(50);
    }
  
  else if(lastBtnState == 0 && btnState == 1){
    MIDI.sendNoteOff(60, 0, 1);
    delay(150);
  
  }

  lastBtnState = btnState;
  btnState = digitalRead(btnPin2);

  if(lastBtnState == 1 && btnState == 0){
    MIDI.sendNoteOn(70, 127, 1);
    delay(50);
    }
  
  else if(lastBtnState == 0 && btnState == 1){
    MIDI.sendNoteOff(70, 0, 1);
    delay(150);
  
  }
}`

Thank you so much in advance for anyone taking the time to look at this and help me :)

**Comment by lathoub** [comment]
Same as discussion BLE-MIDI [93](https://github.com/lathoub/Arduino-BLE-MIDI/discussions/93)

**Comment by lathoub** [comment]
Here is what ChatGPT came up with (of what you can find if you search the web for `MIDI Arduino Potentiometer`)


Certainly! I'll guide you through creating an Arduino MIDI controller that uses a potentiometer to send MIDI Control Change (CC) messages.

Requirements:
Arduino board (e.g., Arduino Uno, Nano)
MIDI-to-USB interface (or a direct USB-MIDI device)
Potentiometer (10kΩ recommended)
MIDI library for Arduino

Steps:
Set up the hardware:
Connect the potentiometer to your Arduino:

GND pin of the potentiometer to the GND of Arduino.
VCC pin of the potentiometer to the 5V of Arduino.
Signal pin (middle pin) of the potentiometer to an analog input pin on the Arduino (e.g., A0).
###Install the MIDI library:
You'll need the MIDI Library for Arduino. Install it via the Arduino IDE:

Go to Sketch -> Include Library -> Manage Libraries.
Search for "MIDI Library" and install it.

```
#include <MIDI.h>

// Create an instance of the MIDI interface
MIDI_CREATE_DEFAULT_INSTANCE();

// Define the analog pin for the potentiometer
const int potPin = A0;

// Define the CC number you want to send
const int ccNumber = 7;  // CC 7 is usually volume

// Define the MIDI channel (1-16)
const int midiChannel = 1;

// Variable to store the last CC value sent
int lastCcValue = -1;

void setup() {
  // Initialize the MIDI interface
  MIDI.begin(MIDI_CHANNEL_OMNI);  // Listen to all channels
}

void loop() {
  // Read the potentiometer value (0-1023)
  int potValue = analogRead(potPin);

  // Map the value to the MIDI CC range (0-127)
  int ccValue = map(potValue, 0, 1023, 0, 127);

  // Only send the CC value if it has changed
  if (ccValue != lastCcValue) {
    // Send the CC message
    MIDI.sendControlChange(ccNumber, ccValue, midiChannel);

    // Update the last sent value
    lastCcValue = ccValue;
  }

  // Add a small delay to avoid flooding MIDI messages
  delay(10);
}
```

Upload the code:
Connect your Arduino to the computer.
Select the correct board and port in the Arduino IDE.
Upload the code to your Arduino.
Connect to a MIDI interface:
If you're using a MIDI-to-USB interface, connect the Arduino's TX pin to the MIDI interface's RX pin.
If using a direct USB-MIDI connection, you can skip this step.
Test the setup:
Open a MIDI monitoring tool or your DAW (Digital Audio Workstation).
Turn the potentiometer, and you should see the MIDI CC messages being sent.
This code will send MIDI CC messages based on the position of the potentiometer, which you can use to control various parameters in your music software or hardware.
